pkg/steps/cluster_claim.go

```diff
@@ -45,81 +47,91 @@ func (s *clusterClaimStep) Validate() error {
 	if s.clusterClaim == nil {
 		// should never happen
 		return fmt.Errorf("cannot claim a nil cluster")
 	}
+	clusterClaim, err := acquireCluster(ctx, *s.clusterClaim, s.hiveClient, s.client, *s.jobSpec)
+	if err != nil {
+		return err
+	}
+
+	wrappedErr := results.ForReason("executing_test").ForError(s.wrapped.Run(ctx))
+	logrus.Infof("Releasing cluster claims for test %s", s.Name())
+	releaseErr := results.ForReason("releasing_cluster_claim").ForError(releaseCluster(ctx, s.hiveClient, clusterClaim))
+
+	// we want a sensible output error for reporting, so we bubble up these individually
+	//if we can, as this is the only step that can have multiple errors
+	if wrappedErr != nil && releaseErr == nil {
+		return wrappedErr
+	} else if wrappedErr == nil && releaseErr != nil {
+		return releaseErr
+	} else {
+		return utilerrors.NewAggregate([]error{wrappedErr, releaseErr})
+	}
+}
+
+func acquireCluster(ctx context.Context, clusterClaim api.ClusterClaim, hiveClient ctrlruntimeclient.Client, client loggingclient.LoggingClient, jobSpec api.JobSpec) (*hivev1.ClusterClaim, error) {
 	clusterPools := &hivev1.ClusterPoolList{}
 	listOption := ctrlruntimeclient.MatchingLabels{
-		"product":      string(s.clusterClaim.Product),
-		"version":      s.clusterClaim.Version,
-		"architecture": string(s.clusterClaim.Architecture),
-		"cloud":        string(s.clusterClaim.Cloud),
-		"owner":        s.clusterClaim.Owner,
+		"product":      string(clusterClaim.Product),
+		"version":      clusterClaim.Version,
+		"architecture": string(clusterClaim.Architecture),
+		"cloud":        string(clusterClaim.Cloud),
+		"owner":        clusterClaim.Owner,
 	}
-	if err := s.hiveClient.List(ctx, clusterPools, listOption); err != nil {
-		return fmt.Errorf("failed to list cluster pools with list option %v: %w", listOption, err)
+	if err := hiveClient.List(ctx, clusterPools, listOption); err != nil {
+		return nil, fmt.Errorf("failed to list cluster pools with list option %v: %w", listOption, err)
 	}
 
 	l := len(clusterPools.Items)
 	if l == 0 {
-		return fmt.Errorf("failed to find a cluster pool providing the cluster: %v", listOption)
+		return nil, fmt.Errorf("failed to find a cluster pool providing the cluster: %v", listOption)
 	} else if l > 1 {
-		return fmt.Errorf("find %d cluster pools providing the cluster (%v): should be only one", len(clusterPools.Items), listOption)
+		return nil, fmt.Errorf("find %d cluster pools providing the cluster (%v): should be only one", len(clusterPools.Items), listOption)
 	}
 
 	clusterPool := clusterPools.Items[0]
-	claimName := s.jobSpec.ProwJobID
+	claimName := jobSpec.ProwJobID
 	claimNamespace := clusterPool.Namespace
 	// https://github.com/kubernetes/test-infra/blob/d14589b797fae426bb70cea4843fa46be50c6739/prow/pod-utils/decorate/podspec.go#L99-L101
-	jobNameForLabel := s.jobSpec.Job
+	jobNameForLabel := jobSpec.Job
 	if len(jobNameForLabel) > validation.LabelValueMaxLength {
```

Comment:
nit: we have helpers `mungeLabels` and `trimLabels` for this, can we please use them

Reply:
https://github.com/openshift/ci-tools/pull/1934